glib/src/future_with_timeout.rs

```diff
@@ -0,0 +1,95 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use futures_util::{
+    future::{self, Either, Future},
+    pin_mut,
+};
+
+// rustdoc-stripper-ignore-next
+/// The error returned when a future times out.
+pub struct FutureWithTimeoutError;
+
+// rustdoc-stripper-ignore-next
+/// Add a timeout to a `Future`.
+///
+/// The `Future` must be spawned on an `Executor` backed by a `glib::MainContext`.
+pub async fn future_with_timeout_with_priority<T>(
+    priority: crate::Priority,
+    timeout: std::time::Duration,
+    fut: impl Future<Output = T>,
+) -> Result<T, FutureWithTimeoutError> {
+    let timeout = crate::timeout_future_with_priority(priority, timeout);
+    pin_mut!(fut);
+
+    match future::select(fut, timeout).await {
+        Either::Left((x, _)) => Ok(x),
+        _ => Err(FutureWithTimeoutError),
+    }
+}
+
+// rustdoc-stripper-ignore-next
+/// Add a timeout to a `Future`.
+///
+/// The `Future` must be spawned on an `Executor` backed by a `glib::MainContext`.
+pub async fn future_with_timeout<T>(
```

Comment:
@bilelmoussaoui @GuillaumeGomez Any suggestions for a better name?

Reply:
maybe we can have a timeout future type and then have its own constructor rather than having a fn?

```
glib::futures::Timeout::new(timeout: std::time::Duration, fut: impl Future<Output = T>)
```
(I proposed a `futures` module in case we have more, eg `Idle` and `Cancellable`, but this is not a requirement)

Reply:
Having a futures module seems useful, yes. The main question still stays: we have "timeout" as the name for the thing that just triggers at some point, and for the new thing that times out another future.

Reply:
`Timed`?

Reply:
`timeout_future`? Generally it goes like this no?

Reply:
Like https://gtk-rs.org/gtk-rs-core/git/docs/glib/fn.timeout_future.html ? That's exactly the problem :)

Reply:
Damn...